src/jira/client/jira-api-client.ts

```diff
@@ -0,0 +1,494 @@
+import Logger from "bunyan";
+import { getAxiosInstance, JiraClientError } from "./axios";
+import { AxiosInstance, AxiosResponse } from "axios";
+import { Installation } from "models/installation";
+import { Subscription } from "models/subscription";
+import { envVars } from "config/env";
+import { uniq } from "lodash";
+import { jiraIssueKeyParser } from "utils/jira-utils";
+import { TransformedRepositoryId, transformRepositoryId } from "~/src/transforms/transform-repository-id";
+import { getJiraId } from "../util/id";
+import { getCloudOrServerFromGitHubAppId } from "utils/get-cloud-or-server";
+import { getDeploymentDebugInfo, extractDeploymentDataForLoggingPurpose } from "./jira-client-deployment-helper";
+import {
+	truncateIssueKeys,
+	getTruncatedIssueKeys,
+	withinIssueKeyLimit,
+	updateIssueKeyAssociationValuesFor,
+	extractAndHashIssueKeysForLoggingPurpose,
+	safeParseAndHashUnknownIssueKeysForLoggingPurpose,
+	dedupIssueKeys,
+	updateIssueKeysFor,
+	withinIssueKeyAssociationsLimit,
+	truncate
+} from "./jira-client-issue-key-helper";
+import {
+	JiraBuildBulkSubmitData,
+	JiraCommit,
+	JiraDeploymentBulkSubmitData,
+	JiraIssue,
+	JiraSubmitOptions,
+	JiraVulnerabilityBulkSubmitData
+} from "interfaces/jira";
+
+const issueKeyLimitWarning = "Exceeded issue key reference limit. Some issues may not be linked.";
+
+export interface DeploymentsResult {
+	status: number;
+	rejectedDeployments?: any[];
+}
+
+export class JiraClient {
+	axios: AxiosInstance;
+	logger: Logger;
+	jiraHost: string;
+	gitHubInstallationId: number;
+	gitHubAppId: number | undefined;
+
+	static async create(installation: Installation, gitHubAppId: number | undefined, logger: Logger): Promise<JiraClient> {
+		const jiraClient = new JiraClient(installation.jiraHost, installation.id, gitHubAppId, logger);
+		await jiraClient.initialize(installation);
+		return jiraClient;
+	}
+
+	private async initialize(installation: Installation): Promise<void> {
+		const secret = await installation.decrypt("encryptedSharedSecret", this.logger);
+		this.axios = getAxiosInstance(installation.jiraHost, secret, this.logger);
+	}
+
+	private constructor(jiraHost: string, gitHubInstallationId: number, gitHubAppId: number | undefined, logger: Logger) {
```

Comment:
**private constructor** - we need to make a async call to get a secret and since
constructors are a synchronous call as a rule, we get around it with the create function calling the constructor, its private so noone accidentally tries.

its similar to one of the old JCs but its just cleaned up and works for all now.